### PR TITLE
[RLlib] Move DQN into the TargetNetworkAPI (and deprecate `RLModuleWithTargetNetworksInterface`).

### PR DESCRIPTION
**Pull request by sven1977**
Move DQN into the TargetNetworkAPI (and deprecate `RLModuleWithTargetNetworksInterface`).
* Also deprecate `TorchDDPRLModuleWithTargetNetworksInterface`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/appo/appo.py [hunk 1]
```diff
@@ -105,7 +106,7 @@ def __init__(self, algo_class=None):
         self.num_multi_gpu_tower_stacks = 1
         self.minibatch_buffer_size = 1
         self.num_sgd_iter = 1
-        self.target_update_frequency = 1
+        self.target_network_update_freq = 1
```
Comment (simonsays1980) [inline rllib/algorithms/appo/appo.py:109]: Way better!

File: rllib/algorithms/dqn/dqn_rainbow_learner.py [hunk 2]
```diff
@@ -42,12 +46,11 @@ class DQNRainbowLearner(Learner):
...
         super().build()
 
-        # Initially sync target networks (w/ tau=1.0 -> full overwrite).
-        # TODO (sven): Use TargetNetworkAPI as soon as DQN implements it.
+        # Make target networks.
         self.module.foreach_module(
-            lambda mid, module: (
-                module.sync_target_networks(tau=1.0)
-                if hasattr(module, "sync_target_networks")
+            lambda mid, mod: (
```
Comment (simonsays1980) [inline rllib/algorithms/dqn/dqn_rainbow_learner.py:51]: Nice!

File: rllib/core/rl_module/rl_module_with_target_networks_interface.py
Comment (simonsays1980) [inline rllib/core/rl_module/rl_module_with_target_networks_interface.py:1]: And .... gone :)

**Review by simonsays1980 — APPROVED** [review]
LGTM. Can we test, if this still works out with multi-learner setups?

**Comment by sven1977** [comment]
Good point! WE haven't re-merged the +GPU +multi-GPU +multi-agent tests for DQN + SAC yet, correct?

File: rllib/core/rl_module/rl_module_with_target_networks_interface.py
Comment (sven1977) [reply rllib/core/rl_module/rl_module_with_target_networks_interface.py:1]: Yeah, this was such a hack.

**Comment by simonsays1980** [comment]
> Good point! WE haven't re-merged the +GPU +multi-GPU +multi-agent tests for DQN + SAC yet, correct?

No, we haven't because they are flaky on the CI machines.

**Pull request merged**